Add mode check for device

hqhq opened the pull request:
This fixes two problems:
1. `docker run --device /dev/sda:rw ubuntu bash` doesn't work
2. --device /dev/zero:/dev/noro:ro doesn't show clear error message,
   but fail when writing to cgroup file.

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

vdemeester commented:
> `docker run --device /dev/sda:rw ubuntu bash` doesn't work

@hqhq you mean it's not supposed to work, so we should validate it, right ?

vdemeester commented on runconfig/parse_test.go:
```diff
@@ -322,10 +322,10 @@ func TestParseDevice(t *testing.T) {
 			PathInContainer:   "/something",
-			CgroupPermissions: "ro",
+			CgroupPermissions: "rw",
 		},
```
I think you should also add a test for `/dev/snd:r` (and/or `/dev/snd:rwm`, …). :wink:

calavera replied:
Also invalid tests for `rr`, `rwrwrw`, `mmmrwf` and `rf`

hqhq replied:
I think this unit test is for testing if devices option works. For validate part, that's in opts/opts_test.go:TestValidateDevice(), I added two invalid cases about `rr`, is that ok?

vdemeester replied:
Ah right, it's probably better in `TestValidateDevice` :wink:. You could add a few more invalid cases (the one @calavera gave), but other than that :+1:.

cpuguy83 commented on opts/opts.go:
```diff
@@ -170,11 +170,41 @@ func ValidateLink(val string) (string, error) {
+	var i int
+	for i = 0; i < len(mode); i++ {
+		c := mode[i : i+1]
+		if legalDeviceMode[c] {
```
```go
if !legalDeviceMode[c] {
    break
}
legalDeviceMode[c] = false
```

calavera replied:
or you can just rewrite this whole loop using runes:

```go
legalDeviceMode := map[rune]bool{'r': true, 'w': true, 'm': true}
for _, c := range mode {
  if !legalDeviceMode[c] {
    return false
  }
  legalDeviceMode[c] = false
}
return true

cpuguy83 commented:
One nit since there needs to be a change anyway.

calavera commented on opts/opts.go:
```diff
@@ -170,11 +170,41 @@ func ValidateLink(val string) (string, error) {
+	volumeMode = 2
+)
+
+func ValidateDeviceMode(mode string) bool {
```
This function is going to need godoc.

calavera commented on opts/opts.go:
```diff
@@ -182,27 +212,33 @@ func ValidateDevice(val string) (string, error) {
 }
 
-func validatePath(val string, validateMountMode bool) (string, error) {
+func validatePath(val string, validateMode int) (string, error) {
```
why don't you pass the validator function here instead of an int? It can be cleaner if we define this as:

```go
func validatePath(val string, validator func(string) (bool, error)) (string, error)
```

And then you call the validator directly:

```go
validator(mode)
```

hqhq replied:
Because they don't have the same return value requirements for different validation functions, and it might be more flexible if we just pass an int value for validate type and let the function decide what to do, maybe it'll be simple enough to not call a validate function. WDYT?

calavera replied:
I think that passing int values to control flow is a bad design decision, but it's just my personal opinion. I'm sure we can figure out how to have a common interface for both validators.

vdemeester replied:
I tend to agree with @calavera, it would be better passing the validator function. 

> Because they don't have the same return value requirements for different validation functions, and it might be more flexible if we just pass an int value for validate type and let the function decide what to do, maybe it'll be simple enough to not call a validate function. WDYT?

We are already doing that for other validator function (with this `ValidatePath` function — [ValidatorFctType](https://github.com/docker/docker/blob/master/opts/opts.go#L149)) ; The `validateDeviceMode` would return `bool, nil` for all cases — that's a small tradeoff for a cool looking method :sweat_smile:.

But like @calavera this is also my personal opinion.

hqhq replied:
@calavera @vdemeester I agree pass validator function is better, so I added a commit to change `volume.ValidateMountMode` a bit, what about now?

vdemeester replied:
I think it looks good :relaxed:.

hqhq commented:
@vdemeester @cpuguy83 @calavera Updated, thanks.

hqhq committed:
Change return value for ValidateMountMode

1. rename it from ValidateMountMode to ValidMountMode
Because it's a function simply check mount mode is valid or not.
2. remove the rw check return value
It's not supposed to be combined into this function, and we already
have a function for that check.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

hqhq committed:
Add mode check for device

This fixes two problems:
1. docker run --device /dev/sda:rw ubuntu bash doesn't work
2. --device /dev/zero:/dev/noro:ro doesn't show clear error message,
but fail when writing to cgroup file.

Signed-off-by: Qiang Huang <h.huangqiang@huawei.com>

cpuguy83 commented:
LGTM

calavera commented:
do we need docs for this?

cpuguy83 commented:
Doesn't seem like it would need docs to me.

LK4D4 commented:
LGTM
